### ISSUE TITLE
Environmental Sounds

### ISSUE THREAD

**Issue by rubenwardy** [minetest/minetest#2272]
There are solutions to this, but they are slow.

I'd like your suggestions on how you would implement this and any things you'd like it to support.

# Sounds from nodes
* Constant - water waves
* Start/Stop - Lua could enable or disable a sound on a node. Say turning on/off a machine. But this could just use different nodes
* Random - bees, occasional russling
* Collision - walking through

# Environment Sounds
* Birds, crickets at night time.
* Sounds could depend on day/night.
* Wind

# Client / Server
* It should work on an old server, but mods can declare environment sounds.
* Minimize transfer of sounds - don't download sounds if the client has them

**Comment by indriApollo** [comment]
:+1:

**Comment by C1ffisme** [comment]
:+1:

**Comment by rubenwardy** [comment]
Just a thought, sounds from the environment aren't completely detached from nodes, they should depend on an average of the nodes around you. You're more likely to hear a bird singing in the forest than in a desert. But they don't occur at a node or single type of node. But no particular node sound emit a bird due to it's definition, they can fly and go anywhere.

Ideally this would be combined with a simple bird mob, but that would require c++ support to be fast and efficient.

**Comment by MoNTE48** [comment]
This is realy good idea!

**Comment by kilbith** [comment]
I'm all for it. A good audio experience is kind of important for the immersivity. Lua is too slow for that purprose, thus it cause too often a "desync" between the sound playing and its respective environment.

**Comment by MoNTE48** [comment]
Yes, only C++.Lua very slow.

**Comment by tenplus1** [comment]
I requested an update for the find_nodes_in_area command that would drastically improve the speed of node checking in an area for ambiance mods and help cut back on lag:

https://github.com/minetest/minetest/issues/2326
https://github.com/tenplus1/ambience

**Comment by Dopium** [comment]
+1

**Comment by neoascetic** [comment]
I want it! It is rather ugly to do sounds generation work for each player on server-side.

**Comment by 0-afflatus** [comment]
I would hope to be able to set sounds up something like this, naturally YIMV:
```lua
minetest.register_node("test:alarm", {
    sounds = {
        ambient = {name = "test_alarm_bg.ogg", volume = 50, radius = 10},
        looped = {name = "test_alarm_loop.ogg", volume = 150, radius = 5},
        triggered = {name = "test_alarm.ogg", volume = 255, radius = 10},
    },
    on_punch = function(pos, node, player, pointed_thing)
        node:play_sound("triggered")
    end,
})
```

**Comment by 0-afflatus** [comment]
Ambient noises could then be triggered by abms.
The engine probably needs to be fairly brutal in only triggering the sound with the greatest volume relative to the player at any point in time. The resulting sounds need to be mixed and positioned, but I guess the engine already does that.

**Comment by 0-afflatus** [comment]
I would be looking to set ambient sounds up something like this YIMV:
```lua
minetest.register_node("test:alarm", {
    sounds = {
        ambient = {name = "test_alarm_bg", gain = 0.2, radius = 10},
        looped = {name = "test_alarm_loop", gain = 0.5, radius = 5},
        triggered = {name = "test_alarm", gain = 1.0, radius = 10},
    },
    on_punch = function(pos, node, player, pointed_thing)
        node:play_sound("triggered")
    end,
})
```

Ambient sounds could then be triggered with abms using node:play_sound(). The engine would need to be fairly brutal in only playing the loudest sound relative to the player at any point in time.

**Comment by 0-afflatus** [comment]
@rubenwardy 
> Just a thought, sounds from the environment aren't completely detached from nodes, they should depend on an average of the nodes around you. You're more likely to hear a bird singing in the forest than in a desert. But they don't occur at a node or single type of node. But no particular node sound emit a bird due to it's definition, they can fly and go anywhere.

You would get the right effect by attaching birdsong to leaves nodes with a fairly big radius. Birds sing more at rest and only intermittently, if at all, in flight. Attach cicadas to grass and wolf howls to pine trunks (with a very big radius), plus actual passing mob calls and I think it will sound right. IMO sound needs to be positional, so it stays in context. An averaged mix of sounds contains no information, the player should get the loudest sound relative to their position.

**Comment by Ferk** [comment]
I don't think this is a matter of C++ vs Lua. It's more of a matter of client-side vs server-side.

I don't think this should be done using ABMs or anything like that. A new type of modifier should be added that is exclusively used to change sounds (not running an "action" lua function) and that is handled completely client-side after the modified is registered.

**Comment by rubenwardy** [comment]
I totally agree, a Lua client side mod would probably work fine. It does depend on the LOD though. The problem is that it need to search around to then place sources of sound. Sound effects in video games seems very primative, at least compared to graphics - you're limited by sound sources as too many will overlay, there's also no good support for refraction/echos/doppler.

**Comment by neoascetic** [comment]
+1. I think ability to implement client-side mods is more general solution and also will fix other issues.

**Comment by ObaniGemini** [comment]
Yes. Quake-like games (including Teeworlds Tremulous) usually do it very well.
Also if something is modified in the pk3 in Tremulous, it's redownloaded. (pk3 = zip)

**Comment by sorcerykid** [comment]
I think a decent environmental sound API, particularly in a client-server model, will necessitate a great deal of forethought.

I would imagine that certain sounds can be ambient (such as wind or waves)  while others can be random (generally positional, like birds chirping) or incidental (generally non-positional, like a chest opening). Client-side prediction would therefore be possible, but that would still require server synchronization for the utmost realism. A frog randomly croaking in the distance wouldn't make sense, if only one player hears it and comments "Whoa! Did you hear that?" and the others respond "What are you talking about? I didn't hear a frog."

**Comment by Fixer-007** [comment]
Waiting for client-side lua and then boom.

**Comment by neoascetic** [comment]
@sorcerykid example with a frog is nice, thus there is not need to wait for client-side lua support :)

**Comment by sorcerykid** [comment]
I think client side Lua certainly has a lot of potential, and would be well suited for certain situations. For example footstep sounds (to my knowledge) are already handled entirely by the client. In tandem with client side Lua, however, there would be a wealth of new possibilities for dynamic and interactive soundscapes.

**Comment by Wuzzy2** [comment]
Environmental sounds are very much needed. I am very in favor of rubenwardy's suggestion to base environment sound on the physical nodes near the player. So it's not completely random.
Also the current biome could be taken into account as well.

We already have mods for this but most of them are very inefficient.


Also we need something to play sounds for these scenarios:
* Fire (with engine support instead of hacky laggy mods)
* Flowing water
* The ocean / very large body of water (must be careful not to play ocean sound in water caves)
* Lava lakes

**Comment by nerzhul** [comment]
if i remember biome is not sent to client, it can only be guess

**Comment by sorcerykid** [comment]
Hopefully at some point soon, I will release my environmental sounds mod. It is specifically optimized for use in multiplayer environments and performs much better than all the existing solutions. The API just needs to some more fine tuning, but other than that the underlying logic is completely solid.

**Comment by jordan4ibanez** [comment]
This should be part of the game already, without it, it's lame af.

**Comment by paramat** [comment]
Wut. Go play another game then.

**Comment by nerzhul** [comment]
It should be part of thé game but client side

**Comment by rubenwardy** [comment]
Agreed. The engine does need support for searching nearby nodes and playing sounds for them

**Comment by MarkuBu** [comment]
> The engine does need support for searching nearby nodes and playing sounds for them

I already do this in my ambiance mod. Works very well

https://github.com/MarkuBu/ambianceplus

**Comment by rubenwardy** [comment]
> I already do this in my ambiance mod. Works very well

That's not client side.

**Comment by paramat** [comment]
A useful example of the issues is Taoki's attempt at adding flowing water sounds.

I managed to code fire sounds to be lightweight, i would like to add water sounds to MTG and am wondering if i can do this (temporarily) with current code in a similar way.

**Comment by DS-Minetest** [comment]
@paramat "af" stands for Afghanistan, angiofibroma, autofocus, aerodynamic force, aggressive fibromatosis, amniotic fluid, atrial fibrillation and much more, but not for "as f**k".

**Comment by sorcerykid** [comment]
I guess it might be time for me to finally release my ambience mod. It affords a robust and lightweight server-side alternative to the existing solutions that seem to rely entirely on ABMs. I opted for a heuristic-based approach, since large-scale prediction and intelligent caching of the multiplayer environment is more efficient and realistic than random-firing timers against matching nodes.

And also, I disagree that the optimal solution must be client-side, as I explained my reasons above.

https://github.com/minetest/minetest/issues/2272#issuecomment-277463858

**Comment by paramat** [comment]
Folded #1369 into this.
